File: site/app/detection/batch-prior.ts
import type {
  BatchPrior,
  PreliminaryResult,
  Quad,
  QuadCandidate,
} from "./types.ts";
import { DETECTION_CONFIG } from "./config.ts";
import { createCandidate, emptyCandidateFeatures } from "./candidate-factory.ts";
import { average, clamp, round } from "./numeric.ts";

const BATCH_CONFIG = DETECTION_CONFIG.batchPrior;

export function buildBatchPriors(results: PreliminaryResult[]): BatchPrior[] {
  const reliable = results.filter((result) =>
    result.confidence >= BATCH_CONFIG.minimumReliableConfidence && result.method !== "fallback-frame"
  );
  const clusters: PreliminaryResult[][] = [];
  for (const result of reliable) {
    const orientation = imageOrientation(result.width, result.height);
    let bestCluster: PreliminaryResult[] | null = null;
    let bestDistance = Number.POSITIVE_INFINITY;
    for (const cluster of clusters) {
      if (imageOrientation(cluster[0].width, cluster[0].height) !== orientation) continue;
      const center = medianQuad(cluster.map((item) => item.normalizedQuad));
      const distance = normalizedQuadDistance(result.normalizedQuad, center);
      if (distance < BATCH_CONFIG.clusterDistance && distance < bestDistance) {
        bestCluster = cluster;
        bestDistance = distance;
      }
    }
    if (bestCluster) bestCluster.push(result);
    else clusters.push([result]);
  }

  const priors: BatchPrior[] = [];
  for (const cluster of clusters) {
    if (cluster.length < BATCH_CONFIG.minimumClusterMembers) continue;
    const normalizedQuad = medianQuad(cluster.map((item) => item.normalizedQuad));
    const rmsDeviation = Math.sqrt(
      average(cluster.map((item) => normalizedQuadDistance(item.normalizedQuad, normalizedQuad) ** 2)),
    );
    if (rmsDeviation >= BATCH_CONFIG.maximumRmsDeviation) continue;
    priors.push({
      id: `camera-position-cluster-${priors.length + 1}`,
      orientation: imageOrientation(cluster[0].width, cluster[0].height),
      normalizedQuad,
      memberCount: cluster.length,
      rmsDeviation,
      consistency: clamp(1 - rmsDeviation / BATCH_CONFIG.consistencyScale, 0, 1),
    });
  }
  return priors;
}

export function batchPriorCandidates(
  priors: BatchPrior[],
  width: number,
  height: number,
): QuadCandidate[] {
  const orientation = imageOrientation(width, height);
  return priors
    .filter((prior) => prior.orientation === orientation)
    .map((prior) => createCandidate(
      "batch-prior",
      prior.normalizedQuad.map(([x, y]) => [x * width, y * height]) as Quad,
      0,
      {
        batchPriorId: prior.id,
        batchPriorMemberCount: prior.memberCount,
        batchPriorRmsDeviation: round(prior.rmsDeviation, 5),
        batchPriorConsistency: round(prior.consistency, 4),
      },
      { ...emptyCandidateFeatures(), batchConsistency: prior.consistency },
    ));
}

export function normalizeResult(
  imageId: string,
  width: number,
  height: number,
  quad: Quad,
  confidence: number,
  method: PreliminaryResult["method"],
  needsReview: boolean,
): PreliminaryResult {
  return {
    imageId,
    width,
    height,
    normalizedQuad: quad.map(([x, y]) => [x / width, y / height]) as Quad,
    confidence,
    method,
    needsReview,
  };
}

function medianQuad(quads: Quad[]): Quad {
  return Array.from({ length: 4 }, (_, corner) => [
    median(quads.map((quad) => quad[corner][0])),
    median(quads.map((quad) => quad[corner][1])),
  ]) as Quad;
}

function normalizedQuadDistance(first: Quad, second: Quad): number {
  return average(first.map((point, index) =>
    Math.hypot(point[0] - second[index][0], point[1] - second[index][1])
  ));
}

function imageOrientation(width: number, height: number): "landscape" | "portrait" {
  return width >= height ? "landscape" : "portrait";
}

function median(values: number[]): number {
  const ordered = [...values].sort((first, second) => first - second);
  const middle = Math.floor(ordered.length / 2);
  return ordered.length % 2
    ? ordered[middle]
    : (ordered[middle - 1] + ordered[middle]) / 2;
}

File: src/slides_thief/detection/batch_prior.py
"""Safe multi-image geometry priors for repeated camera positions."""

from __future__ import annotations

import math

import numpy as np

from .candidate_factory import make_candidate
from .config import DETECTION_CONFIG
from .numeric import average

_BATCH_CONFIG = DETECTION_CONFIG["batchPrior"]


def build_batch_priors(results: list[dict]) -> list[dict]:
    reliable = [
        result
        for result in results
        if result["confidence"] >= float(_BATCH_CONFIG["minimumReliableConfidence"])
        and result["method"] != "fallback-frame"
    ]
    clusters: list[list[dict]] = []
    for result in reliable:
        orientation = _orientation(result["width"], result["height"])
        best_cluster = None
        best_distance = math.inf
        for cluster in clusters:
            if _orientation(cluster[0]["width"], cluster[0]["height"]) != orientation:
                continue
            center = _median_quad([item["normalized_quad"] for item in cluster])
            distance = _quad_distance(result["normalized_quad"], center)
            if distance < float(_BATCH_CONFIG["clusterDistance"]) and distance < best_distance:
                best_cluster = cluster
                best_distance = distance
        if best_cluster is None:
            clusters.append([result])
        else:
            best_cluster.append(result)

    priors = []
    for cluster in clusters:
        if len(cluster) < int(_BATCH_CONFIG["minimumClusterMembers"]):
            continue
        normalized_quad = _median_quad([item["normalized_quad"] for item in cluster])
        rms_deviation = math.sqrt(
            average([_quad_distance(item["normalized_quad"], normalized_quad) ** 2 for item in cluster])
        )
        if rms_deviation >= float(_BATCH_CONFIG["maximumRmsDeviation"]):
            continue
        priors.append(
            {
                "id": f"camera-position-cluster-{len(priors) + 1}",
                "orientation": _orientation(cluster[0]["width"], cluster[0]["height"]),
                "normalized_quad": normalized_quad,
                "member_count": len(cluster),
                "rms_deviation": rms_deviation,
                "consistency": min(
                    1.0,
                    max(0.0, 1 - rms_deviation / float(_BATCH_CONFIG["consistencyScale"])),
                ),
            }
        )
    return priors


def batch_prior_candidates(priors: list[dict], width: int, height: int) -> list[dict]:
    orientation = _orientation(width, height)
    candidates = []
    for prior in priors:
        if prior["orientation"] != orientation:
            continue
        scale = np.array([width, height], dtype=np.float64)
        candidates.append(
            make_candidate(
                prior["normalized_quad"] * scale,
                "batch-prior",
                {
                    "batch_prior_id": prior["id"],
                    "batch_prior_member_count": prior["member_count"],
                    "batch_prior_rms_deviation": round(prior["rms_deviation"], 5),
                    "batch_prior_consistency": round(prior["consistency"], 4),
                },
                batch_consistency=prior["consistency"],
            )
        )
    return candidates


def normalize_result(
    image_id: str,
    width: int,
    height: int,
    quad: np.ndarray,
    diagnostics: dict,
) -> dict:
    return {
        "image_id": image_id,
        "width": width,
        "height": height,
        "normalized_quad": quad / np.array([width, height], dtype=np.float64),
        "confidence": diagnostics["confidence"],
        "method": diagnostics["method"],
        "needs_review": diagnostics["needs_review"],
    }


def _median_quad(quads: list[np.ndarray]) -> np.ndarray:
    return np.median(np.stack(quads), axis=0)


def _quad_distance(first: np.ndarray, second: np.ndarray) -> float:
    return float(np.linalg.norm(first - second, axis=1).mean())


def _orientation(width: int, height: int) -> str:
    return "landscape" if width >= height else "portrait"
